data/transactions/logic/eval.go

```diff
@@ -4492,40 +4502,63 @@ func opAppGlobalDel(cx *EvalContext) error {
 // more than 2 or so, and was often called an "index".  But it was not a
 // basics.AssetIndex or basics.ApplicationIndex.
 
-func (cx *EvalContext) appReference(ref uint64, foreign bool) (basics.AppIndex, error) {
+func (cx *EvalContext) appReference(ref uint64, foreign bool) (aid basics.AppIndex, err error) {
 	if cx.version >= directRefEnabledVersion {
-		if ref == 0 || ref == uint64(cx.appID) {
-			return cx.appID, nil
-		}
-		aid := basics.AppIndex(ref)
-		if cx.availableApp(aid) {
-			return aid, nil
-		}
+		return cx.resolveApp(ref)
+	}
 
-		if cx.version < sharedResourcesVersion {
-			// Allow use of indexes, but this comes last so that clear advice can be
-			// given to anyone who cares about semantics in the first few rounds of
-			// a new network - don't use indexes for references, use the App ID
-			if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-				return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+	// resolveApp is already similarly protected (and must be, since it is
+	// called independently)
+	if cx.Proto.AppForbidLowResources {
+		defer func() {
+			if aid <= lastForbiddenResource && err == nil {
+				err = fmt.Errorf("low App lookup %d", aid)
 			}
+		}()
+	}
+	// Old rules, pre directRefEnabledVersion, when a ref has to be a slot for
+	// some opcodes, and had to be an ID for others.
+	if ref == 0 { // Even back when expected to be a real ID, ref = 0 was current app
+		return cx.appID, nil
+	}
+	if foreign {
+		// In old versions, a foreign reference must be an index in ForeignApps or 0
+		if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
+			return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
 		}
-	} else {
-		// Old rules
-		if ref == 0 { // Even back when expected to be a real ID, ref = 0 was current app
-			return cx.appID, nil
-		}
-		if foreign {
-			// In old versions, a foreign reference must be an index in ForeignAssets or 0
-			if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-				return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+		return 0, fmt.Errorf("App index %d beyond txn.ForeignApps", ref)
+	}
+	// Otherwise it's direct
+	return basics.AppIndex(ref), nil
```

Comment:
Which line? 4532?

It would be hit in pre directRefEnabled version usage of `app_opted_in` or `app_local_get`. They require the use of explicit IDs. Slots are not allowed (that's the `foreign` flag in this function, false means "not a slot reference" - maybe not a great name for the flag)

Maybe `directRefEnabled` isn't even the best name. The reality is that is enabled "direct" (ID) based usage on some opcodes, but it _also_ (for symmetry) began allowing slot based usage on some opcodes that previous required IDs.  Before that version, some opcodes required a slot, and some required an ID.

Reply:
> Which line? 4532?

yea line 4532, which is the last else branch of previous impl of `appReference` under `cx.version < directRefEnabledVersion`.

> app_opted_in or app_local_get. They required the use of explicit IDs.

I see, that explains everything then. That also explains why the `foreign` flag separates a case into two.

> Maybe directRefEnabled isn't even the best name.

Actually I have no opinions on its naming, the whole thing I put this question is that, I was not aware `app_opted_in` and `app_local_get` behave under `!foreign` case.